### PR TITLE
Fix Servlet instrumentation for servlet-api 3.1+ - alternative take

### LINKED ISSUE #1083: Servlet 3.0 instrumentation does not fully cover 3.1 and 4.0
Servlet 3.1 introduced a new method in `ServletOutputStream`: `setWriteListener(WriteListener)`. This method is not implemented in our current servlet 3.0 instrumentation, so when it's called an `AbstractMethodError` is thrown instead:

```
 org.jboss.resteasy.spi.UnhandledException: java.lang.AbstractMethodError: Receiver class io.opentelemetry.instrumentation.auto.servlet.v3_0.CountingHttpServletResponse$CountingServletOutputStream does not define or inherit an implementation of the resolved method 'abstract void setWriteListener(javax.servlet.WriteListener)' of abstract class javax.servlet.ServletOutputStream.
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:381)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:216)
	at org.jboss.resteasy.core.SynchronousDispatcher.asynchronousExceptionDelivery(SynchronousDispatcher.java:571)
...
Caused by: java.lang.AbstractMethodError: Receiver class io.opentelemetry.instrumentation.auto.servlet.v3_0.CountingHttpServletResponse$CountingServletOutputStream does not define or inherit an implementation of the resolved method 'abstract void setWriteListener(javax.servlet.WriteListener)' of abstract class javax.servlet.ServletOutputStream.
	at org.jboss.resteasy.plugins.server.servlet.HttpServletResponseWrapper$DeferredOutputStream.queue(HttpServletResponseWrapper.java:204)
	at org.jboss.resteasy.plugins.server.servlet.HttpServletResponseWrapper$DeferredOutputStream.asyncWrite(HttpServletResponseWrapper.java:179)
	at org.jboss.resteasy.util.CommitHeaderAsyncOutputStream.asyncWrite(CommitHeaderAsyncOutputStream.java:94)
	at org.jboss.resteasy.spi.AsyncOutputStream.asyncWrite(AsyncOutputStream.java:23)
...
```

I ran into this issue when I was unit testing RESTEasy 4.5.6: it looks like they actually use this new method when dealing with an `AsyncResponse`: https://github.com/resteasy/Resteasy/blob/master/resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java#L204
This issue is easy to reproduce by using code from [this PR](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/1082/files) and running

```sh
./gradlew :instrumentation:jaxrs:jaxrs-2.0:jaxrs-2.0-resteasy-3.1:test -PtestLatestDeps=true
```

The worst thing about this is that it affects the instrumented application (because we overwrite request & response with our wrappers) - in my case, I got a 500 HTTP response instead of 1 200 one:

```
Response{protocol=http/1.1, code=500, message=Internal Server Error, url=http://localhost:56412/async?action=succeed}
```

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
Fixes #1083

Alternative solution to  #1104 -- does not require as much copy-paste but it's a bit more hacky. No need to change any gradle modules as well.

### CODE REVIEW

File: instrumentation/servlet/servlet-3.0/src/main/java/io/opentelemetry/instrumentation/auto/servlet/v3_0/CountingHttpServletResponse.java [hunk 1]
```diff
@@ -110,6 +113,30 @@ public void flush() throws IOException {
...
     }
+
+    // New abstract methods introduced in servlet-api 3.1, 3.0 does not have them.
+    // These methods purposefully use reflection to call the delegate object so that they compile
+    // against both 3.0 and 3.1+ servlet-api.
+
+    // @Override
+    public boolean isReady() {
+      try {
+        Method isReady = ServletOutputStream.class.getMethod("isReady");
```
Comment (iNikem) [inline instrumentation/servlet/servlet-3.0/src/main/java/io/opentelemetry/instrumentation/auto/servlet/v3_0/CountingHttpServletResponse.java:124]: This certainly should be cached, no need to lookup on every method call. Also, probably `MethodHandle` will work faster than `Method`

File: instrumentation/servlet/servlet-3.0/src/main/java/javax/servlet/WriteListener.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+import java.io.IOException;
+import java.util.EventListener;
+
+/**
+ * Backport of a callback interface introduced in Servlet 3.1 API. Having it here is required to
+ * make {@code Servlet3Instrumentation} work with Servlet 3.0 API, otherwise muzzle would disable it
+ * because of a missing class.
+ */
+public interface WriteListener extends EventListener {
```
Comment (iNikem) [inline instrumentation/servlet/servlet-3.0/src/main/java/javax/servlet/WriteListener.java:27]: And what will happen if there is the original `WriteListener` on the classpath?

**Review by iNikem — COMMENTED** [review]
This is certainly much nicer PR than the other one :) 4 classes vs 45 :)

File: instrumentation/servlet/servlet-3.0/src/main/java/javax/servlet/WriteListener.java (hunk 2 above)
Comment (mateuszrzeszutek) [reply instrumentation/servlet/servlet-3.0/src/main/java/javax/servlet/WriteListener.java:27]: First one found will be loaded, I guess. This class does not change in any [3.1, 5) version, so we shouldn't worry about any conflict or missing methods.

File: instrumentation/servlet/servlet-3.0/src/main/java/io/opentelemetry/instrumentation/auto/servlet/v3_0/CountingHttpServletResponse.java (hunk 1 above)
Comment (mateuszrzeszutek) [reply instrumentation/servlet/servlet-3.0/src/main/java/io/opentelemetry/instrumentation/auto/servlet/v3_0/CountingHttpServletResponse.java]: Very good point, done. Thanks!

**Comment by anuraaga** [comment]
Agreed :) Judging from the CI it seems muzzle may not be liking this yet but it's probably our job to fix muzzle for this sort of use case. Can't dig too deep today but thanks a lot for bearing with this!

**Comment by mateuszrzeszutek** [comment]
Looks like I was too optimistic about this solution: `java.lang.SecurityException: class "javax.servlet.WriteListener"'s signer information does not match signer information of other classes in the same package` -- it seems that just copying a single class to our project is not viable. (It's a bit strange that muzzle check passed on my laptop, but I guess that it may depend on Java version that's being used)

I have several ideas on how to solve this:
1. Go back to the copy-paste solution;
2. Implement a way to make muzzle ignore a selected mismatch, perhaps annotation on method param?;
3. Create a dynamic proxy in runtime -- since we don't care about these new methods at all. This one could be significantly slower than other solutions;
4. Create a dynamic class extending `ServletOutputStream` when instrumentation is loaded -- this should have better performance than 3 bun can be tricky.

What do you all think? I'd be grateful for any other ideas/comments/feedback.

**Pull request closed without merging**